src/modulefinder/sentry_modulefinder_linux.c

```diff
@@ -293,7 +272,33 @@ try_append_module(sentry_value_t modules, const sentry_module_t *module)
         return;
     }
 
-    sentry_value_append(modules, sentry__procmaps_module_to_value(module));
+    // At least on the android API-16, x86 simulator, the linker apparently
+    // does not load the complete file into memory. Or at least, the section
+    // headers which are located at the end of the file are not loaded, and
+    // we would be poking into invalid memory. To be safe, we load the file
+    // from disk, so we have the on-disk layout, and are independent of how the
+    // runtime linker would load or re-order any sections. The exception here is
+    // the linux-gate, which is not an actual file on disk, so we actually poke
+    // at its memory.
+    sentry_value_t mod = sentry_value_new_null();
+
+    if (sentry__slice_eq(module->file, LINUX_GATE)) {
+        mod = sentry__procmaps_module_to_value(module, module->start);
+    } else {
+        char *filename = sentry__slice_to_owned(module->file);
+        sentry_path_t *file = sentry__path_from_str_owned(filename);
+        if (file) {
+            char *buf = sentry__path_read_to_buffer(file, NULL);
+            sentry__path_free(file);
+            if (buf) {
+                mod = sentry__procmaps_module_to_value(module, buf);
```

Comment:
Could create a fake `sentry_module_t` structure here instead of introducing an `addr` parameter? Also, this is potentially unsafe, since the `module->end` pointer no longer fits the address. Probably something like this would be better:

```c
            size_t buf_len;
            char *buf = sentry__path_read_to_buffer(file, &buf_len);
            sentry__path_free(file);
            if (buf) {
                sentry_module_t loaded_module = { buf, buf + buf_len, module->file };
                mod = sentry__procmaps_module_to_value(loaded_module);
```

Reply:
I thought about this, but the `start`/`end` pairs are actually used and put into the resulting `value_t`. Not sure about the naming, but we need to have two different things, one for the module as it is really loaded, and one just for the elf parsing -_-

Reply:
maybe have one function that populated a `sentry_value_t` from a raw pointer to an ELF file.

Reply:
I see, thanks for clarifying. Let's keep your version then. 👍